yapapi/engine.py

```diff
@@ -448,26 +455,48 @@ async def _process_invoices(self) -> None:
             ):
                 break
 
+    async def _check_debit_note_rate(self, act_id: ActivityId, agr_id: AgreementId, job: "Job"):
+        agreement = self._get_agreement_by_id(agr_id)
+        self._number_of_debit_notes[act_id] += 1
+        num_notes = self._number_of_debit_notes[act_id]
+        ts = datetime.now()
+        start_ts = self._activity_created_at.get(act_id)
+        max_interval = job.agreements_pool.max_debit_note_interval_for_agreement(agr_id)
+        if start_ts is not None and max_interval is not None:
+            dur = (ts - start_ts).total_seconds()
+            freq_descr = f"{num_notes} notes/{dur}s"
+            logger.debug("Debit notes for activity {act_id}: {freq_descr}")
+            if dur > 0 and dur < num_notes * max_interval:
+                reason = {
+                    "message": f"Too many debit notes: {freq_descr} (activity: {act_id})",
+                    "golem.requestor.code": "TooManyDebitNotes",
+                }
+                job.emit(events.PaymentFailed, agreement=agreement)
+                await agreement.terminate(reason)
```

Comment:
I'm pretty sure we should call `AgreementsPool._terminate_agreement(agr_id, reason)` here.
It will **hopefully** (i.e. we should check this) do all the stuff necessary to keep AgreementsPool working (e.g. so that it will know this is a dead agreement), and also emit all events necessary.

Reply:
Fixed